File: apps/android/app/src/main/kotlin/com/loonext/android/features/shell/RootViewModel.kt
package com.loonext.android.features.shell

import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleEventObserver
import androidx.lifecycle.ProcessLifecycleOwner
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.loonext.android.AppGraph
import com.loonext.android.core.i18n.AppStrings
import com.loonext.android.core.i18n.UiLocale
import com.loonext.android.core.model.Me
import com.loonext.android.core.model.MemberRole
import com.loonext.android.core.model.SubscriptionStatus
import com.loonext.android.core.net.ApiErrorCode
import com.loonext.android.core.net.ApiException
import com.loonext.android.features.settings.SettingsRepository
import com.loonext.android.core.realtime.RealtimeLifecycle
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch

/**
 * Mirrors the web's CompanyProvider bootstrap: session → GET /v1/me →
 * resolve active company (persisted pick or first membership) → route.
 */
sealed interface RootState {
    data object Loading : RootState
    data object SignedOut : RootState

    /** Signed in, zero memberships — workspace creation lives on web (checkout). */
    data class NeedsWorkspace(val me: Me) : RootState

    /** Owner/admin with subscription_status incomplete — finish checkout on web. */
    data class NeedsCheckout(val me: Me, val companyId: String) : RootState

    /**
     * #496/#314: this session has not satisfied two-factor. `enrolmentRequired`
     * tells the gate which question to ask — a code from somebody who has a
     * factor, or enrolment first for somebody the WORKSPACE is insisting on and
     * who has none.
     */
    data class NeedsMfa(val enrolmentRequired: Boolean) : RootState

    data class Ready(val me: Me, val companyId: String) : RootState

    data class Failed(val message: String) : RootState
}

/**
 * #483: the waits before each retry of a bootstrap number list that failed.
 *
 * The hydrated `/v1/me` is the ONLY source of the access-filtered number list,
 * and when it fails [RootViewModel] hands realtime an empty one — the member
 * holds the company topic and not a single per-number topic. The reconnect
 * collector heals that on the next re-JOIN, which on a healthy socket can be
 * hours away, and since #484's contract step those are hours of an inbox that
 * never updates. Three tries across ~17s turn one transient 5xx into a
 * blink; going longer would only race the re-JOIN heal that already exists.
 */
internal val NUMBER_LIST_RETRY_DELAYS_MS = listOf(1_000L, 4_000L, 12_000L)

/**
 * Run [attempt] after each wait in [NUMBER_LIST_RETRY_DELAYS_MS] until one
 * reports success. Waits FIRST: the read that just failed fails the same way if
 * it is repeated in the same millisecond.
 */
internal suspend fun retryNumberList(attempt: suspend () -> Boolean): Boolean {
    for (delayMs in NUMBER_LIST_RETRY_DELAYS_MS) {
        delay(delayMs)
        if (attempt()) return true
    }
    return false
}

class RootViewModel(private val graph: AppGraph) : ViewModel() {
    private val _state = MutableStateFlow<RootState>(RootState.Loading)
    val state: StateFlow<RootState> = _state

    /** #483: the in-flight [retryNumberList], so bootstraps cannot stack them. */
    private var numberListRetry: Job? = null

    /**
     * #289: the pending background drop, so a quick app-switch cancels it
     * rather than stacking a second one behind it.
     */
    private var realtimeDrop: Job? = null

    /**
     * #289: whether the socket is down because the app is backgrounded, as
     * opposed to never having been up.
     *
     * Needed because ON_START fires on every return to the app, and reconnecting
     * a socket that is already connected would tear down a live one for nothing.
     */
    private var droppedForBackground = false

    init {
        // Session appearing/disappearing drives everything.
        viewModelScope.launch {
            graph.sessionStore.session.collect { session ->
                if (session == null) {
                    graph.realtime.disconnect()
                    _state.value = RootState.SignedOut
                } else if (_state.value is RootState.SignedOut ||
                    _state.value is RootState.Loading
                ) {
                    bootstrap()
                }
            }
        }
        // #289: the socket a backgrounded phone should not be holding.
        //
        // Both apps connected on sign-in and disconnected on sign-out, so a
        // phone in a pocket kept a WebSocket alive and sent a heartbeat every
        // 25 seconds all day. The bytes are nothing; the RADIO is the cost —
        // on LTE each transmission holds the modem in a high-power state for
        // seconds afterwards, so a packet every 25 seconds never lets it sleep.
        //
        // Nothing is lost by dropping it: push carries every message, task and
        // — the unforgivable one — every incoming call, which is woken through
        // PushHooks.callWakeHandler and never by a socket frame.
        watchAppVisibility()

        // A dead refresh token anywhere lands back on login.
        viewModelScope.launch {
            graph.api.signedOut.collect {
                graph.realtime.disconnect()
                _state.value = RootState.SignedOut
            }
        }
        // #480: somebody's number access changed in this company. Which numbers
        // this member may see is a SERVER answer and it may just have changed, so
        // ask again and move the per-number subscriptions with it.
        //
        // Re-subscribing is the part a plain refetch cannot do. Realtime
        // authorization is a join-time handshake (D88 addendum), so a member who
        // just lost a number stays joined to its topic and keeps receiving its
        // events until their JWT refreshes — up to an hour of a boundary the
        // product believes it is enforcing. The event names only the company, so
        // a client cannot tell whether it was the subject; it just asks again.
        viewModelScope.launch {
            graph.realtime.events.collect { event ->
                // LAUNCHED, not awaited. reconnectRealtime() makes two network
                // calls, and this collector shares one MutableSharedFlow whose
                // overflow policy is SUSPEND at 64 buffered frames: the single
                // dispatch pump stalls once the SLOWEST collector falls behind,
                // which would starve every other collector in the app — inbox,
                // thread, calls, tasks, notifications — of every frame while this
                // one waited on /v1/me behind a captive portal.
                if (event.event == "access.changed") {
                    viewModelScope.launch { reconnectRealtime() }
                }
            }
        }

        // #480: broadcasts are NOT replayed. An access.changed published while
        // this app was backgrounded or out of signal is gone, so on reconnect we
        // would re-join exactly the pre-gap topic set — a number granted during
        // the outage would have no realtime for the rest of the process lifetime,
        // with the socket reporting Joined the whole time.
        //
        // Re-deriving on every re-JOIN is the heal. It also covers a number list
        // that failed to load at bootstrap, which otherwise left the member with
        // no per-number subscriptions at all until the app was killed.
        //
        // AWAITED inside the callback, unlike the `access.changed` collector
        // above, which cannot be. Two round trips in here no longer cost anyone an
        // edge: `reconnected`'s overflow policy is DROP_OLDEST (#483), so a signal
        // arriving while this works replaces the pending one and is delivered when
        // it returns, and the eleven screen collectors are handed it meanwhile.
        // Launching per signal would instead put two /v1/me-then-connect passes in
        // flight at once, and the loser of that race hands the realtime client the
        // OLDER number list — leaving a revoked number joined, or a granted one
        // unjoined, until some later signal. Serial and one behind beats parallel
        // and wrong.
        viewModelScope.launch {
            graph.realtime.reconnected.collect { reconnectRealtime() }
        }
    }

    fun retry() {
        _state.value = RootState.Loading
        viewModelScope.launch { bootstrap() }
    }

    fun switchWorkspace(companyId: String) {
        viewModelScope.launch {
            graph.prefs.setActiveCompany(companyId)
            bootstrap()
        }
    }

    fun signOut() {
        viewModelScope.launch {
            // Drop the device push token while the bearer still works —
            // best-effort; a dead token also self-prunes server-side (#151).
            runCatching { graph.pushRegistrar.unregister() }
            // Revoke the session SERVER-side, in the same "while the bearer still
            // works" slot and for a sharper reason. `AuthManager.signOut` ends the
            // GoTrue session and nothing else, so `user_sessions.revoked_at` stayed
            // null and the softphone credential was never swept — and because
            // authorization is `revoked_at is null`, the access token this phone is
            // holding kept full read and send for the rest of its life after
            // somebody pressed Sign out.
            //
            // Best-effort, like the line above and for the same reason: nobody may
            // be trapped in an account because the network blipped on the way out.
            // The cost of a failure is that the token expires on its own instead of
            // being cut short, which is the state we were in for every sign-out.
            runCatching { SettingsRepository(graph.api).revokeThisSession() }
            graph.authManager.signOut()
        }
    }

    /**
     * Re-derive the visible numbers and hand the new set to the realtime client,
     * which joins what was added and LEAVES what was taken away.
     *
     * Silent on failure by design: this runs off a broadcast, not off something
     * the member did, so a failed read must leave the shell exactly as it was.
     * The subscriptions are then a moment stale until the next reconnect, and a
     * topic we no longer have access to is refused when we next try to join it.
     *
     * Reports whether the list actually moved, which is what [retryNumberList]
     * needs to know when it is standing in for a bootstrap read that failed
     * (#483). The two collectors above discard it: they fire off a signal that
     * will come again, so there is nothing for them to do about a `false`.
     */
    /**
     * #289: drop the socket when the phone goes in a pocket; bring it back when
     * somebody looks at the app.
     *
     * The grace window is what makes this safe to do at all — see
     * [RealtimeLifecycle.BACKGROUND_GRACE_MS]. A live call holds the socket
     * regardless: call state rides realtime, and a call is exactly when the
     * phone is out and being used.
     */
    private fun watchAppVisibility() {
        ProcessLifecycleOwner.get().lifecycle.addObserver(
            LifecycleEventObserver { _, event ->
                when (event) {
                    Lifecycle.Event.ON_START -> {
                        realtimeDrop?.cancel()
                        realtimeDrop = null
                        if (droppedForBackground) {
                            droppedForBackground = false
                            // Not `connect` directly: the token may have expired
                            // while the app was away, and the number list may
                            // have changed. This is the same path the #480
                            // access-change signal uses.
                            viewModelScope.launch { reconnectRealtime() }
                        }
                    }

                    Lifecycle.Event.ON_STOP -> {
                        realtimeDrop?.cancel()
                        realtimeDrop = viewModelScope.launch {
                            val wait = RealtimeLifecycle.dropDelayMs(
                                foreground = false,
                                backgroundedForMs = 0,
                                callActive = graph.callActive(),
                            ) ?: return@launch
                            delay(wait)
                            // Asked again after the wait: a call can start
                            // inside the grace window (a push wakes the app,
                            // rings, and the person answers from the lock
                            // screen without the app ever coming forward).
                            if (graph.callActive()) return@launch
                            droppedForBackground = true
                            graph.realtime.disconnect()
                        }
                    }

                    else -> Unit
                }
            },
        )
    }

    private suspend fun reconnectRealtime(): Boolean {
        val ready = _state.value as? RootState.Ready ?: return false
        val refreshed = runCatching { graph.meRepo.me(ready.companyId) }.getOrNull() ?: return false
        // #228: this runs on every return to the foreground, so it is where a
        // language changed on the laptop reaches the phone.
        graph.prefs.setUiLocale(refreshed.locale)
        graph.prefs.setWorkspaceLocale(
            refreshed.memberships.firstOrNull { it.company_id == ready.companyId }?.locale,
        )
        val session = graph.sessionStore.session.first() ?: return false
        graph.realtime.connect(
            ready.companyId,
            session.accessToken,
            refreshed.company?.numbers?.map { it.id }.orEmpty(),
            // #302: the presence key. Nothing else on this socket says anything
            // about who the viewer is.
            userId = refreshed.user_id,
        )
        // The screens read the number list off this same `me` (the dial-from
        // picker, the composer's From), so a member who just lost a number stops
        // being offered it here rather than on their next app launch.
        _state.value = RootState.Ready(refreshed, ready.companyId)
        return true
    }

    private suspend fun bootstrap() {
        try {
            val me = graph.meRepo.me()
            // #228: the server has just answered the first two thirds of "what
            // language is this app in", so write both down before anything else
            // uses them. The mirror is what makes the NEXT cold start paint in
            // the right language on its first frame instead of flashing English
            // at a French crew while this call is in flight.
            graph.prefs.setUiLocale(me.locale)
            if (me.memberships.isEmpty()) {
                _state.value = RootState.NeedsWorkspace(me)
                return
            }
            val stored = graph.prefs.currentCompanyId()
            val membership = me.memberships.firstOrNull { it.company_id == stored }
                ?: me.memberships.first()
            graph.prefs.setActiveCompany(membership.company_id)
            // The WORKSPACE half, and it has to wait for the line above: which
            // company's language matters is a question the active membership
            // answers, and somebody in two workspaces has two answers.
            graph.prefs.setWorkspaceLocale(membership.locale)

            val incomplete = membership.subscription_status == SubscriptionStatus.INCOMPLETE ||
                membership.subscription_status == SubscriptionStatus.INCOMPLETE_EXPIRED
            if (incomplete && MemberRole.atLeast(membership.role, MemberRole.ADMIN)) {
                _state.value = RootState.NeedsCheckout(me, membership.company_id)
                return
            }

            // The access-filtered number list and the client flags exist ONLY on
            // the hydrated /v1/me — the one that carries X-Company-Id. The call
            // above cannot ask for either: it is what discovers which company to
            // ask about.
            //
            // Two things ride on this second read. #480: the per-number realtime
            // topics must be joined for exactly the numbers this member may see,
            // and the server is the only thing that knows which those are
            // (`loadCompanyView` filters them, and re-deriving the rule here
            // would be a second implementation of D88). And #283's realtime kill
            // switch, which has never actually worked on Android: the route
            // populates `flags` only when hydrating, so the unhydrated `me()`
            // always reported an empty map, which reads as ON.
            //
            // Best-effort, because a failed hydration must not cost the shell.
            // Without it the member joins the company topic alone, which for the
            // length of D88's dual-publish transition still carries every event.
            // Kept as a nullable rather than folded into `view`, because whether
            // it failed is what decides the #483 retry at the bottom.
            //
            // #314/#496: WHY the failure code is kept and not just the null. A
            // refused hydration and a failed one look identical from a null, and
            // one of them is a wall this app is supposed to render a way through
            // rather than shrug at.
            val hydration = runCatching { graph.meRepo.me(membership.company_id) }
            val hydrated = hydration.getOrNull()
            val refusal = (hydration.exceptionOrNull() as? ApiException)?.code
            val view = hydrated ?: me

            // Connect realtime for the active workspace.
            //
            // #283: unless the realtime kill switch is off. This is the one
            // switch the server cannot enforce — the app holds its own Supabase
            // token and opens its own socket, so there is nothing for the
            // Worker to refuse. Not connecting leaves the app on its normal
            // refresh path: slower, never wrong.
            //
            // `!= false` rather than a truthiness check: an absent flag means
            // "no statement", which must read as ON.
            val realtimeAllowed = view.flags["kill:realtime"] != false
            val session = graph.sessionStore.session.first()
            if (session != null && realtimeAllowed) {
                graph.realtime.connect(
                    membership.company_id,
                    session.accessToken,
                    // Empty is a real state when the read SUCCEEDED: a member
                    // restricted out of every number joins the company topic and
                    // nothing else, and everything company-wide still reaches
                    // them. It is a failure when it came from the unhydrated `me`,
                    // whose `company` is always null — hence the retry below.
                    view.company?.numbers?.map { it.id }.orEmpty(),
                    userId = view.user_id, // #302: the presence key
                )
            }
            // #496/#314: two-factor, which the server has just answered about
            // in the only two ways it can.
            //
            // The CHALLENGE case is asked positively rather than read off the
            // refusal, because there is a window where the server is not
            // refusing yet — a Worker deployed ahead of the migration reports no
            // enrolment — and the app should still ask. /v1/mfa is
            // company-exempt, so it answers even for a session every other route
            // is refusing, which is exactly the session that needs the answer.
            if (refusal == ApiErrorCode.MFA_CHALLENGE_REQUIRED) {
                _state.value = RootState.NeedsMfa(enrolmentRequired = false)
                return
            }
            if (refusal == ApiErrorCode.MFA_REQUIRED) {
                _state.value = RootState.NeedsMfa(enrolmentRequired = true)
                return
            }
            val mfa = runCatching { graph.settingsRepo.mfa() }.getOrNull()
            if (mfa != null && mfa.enrolled && mfa.aal != "aal2") {
                _state.value = RootState.NeedsMfa(enrolmentRequired = false)
                return
            }

            _state.value = RootState.Ready(view, membership.company_id)

            // #483: one transient 5xx on the hydrated read must not cost a whole
            // session of per-number realtime. Retried OFF the bootstrap path —
            // the shell is already Ready above, so nothing here is waited on —
            // and through [reconnectRealtime], which is exactly the same work the
            // `access.changed` and reconnect collectors do.
            //
            // A retry still pending from an earlier bootstrap is stale whatever
            // happened here: this pass has just read for the company it would
            // read for.
            numberListRetry?.cancel()
            numberListRetry = if (hydrated == null) {
                viewModelScope.launch { retryNumberList { reconnectRealtime() } }
            } else {
                null
            }
        } catch (cause: ApiException) {
            if (cause.code == ApiErrorCode.UNAUTHORIZED) {
                _state.value = RootState.SignedOut
            } else {
                _state.value = RootState.Failed(cause.message)
            }
        } catch (cause: Exception) {
            _state.value = RootState.Failed(
                AppStrings.translate(uiLocale(), "shell.bootstrapFailed"),
            )
        }
    }

    /**
     * #228: the reader's language, for the one sentence on this path that is
     * OURS.
     *
     * Every other [RootState.Failed] above carries `cause.message` — the API's
     * own words, which are the API's to translate; a client-side catalogue entry
     * for a server refusal would be a second copy that drifts. This one is the
     * catch-all we wrote, and there is no composition here to read a locale from,
     * so it is resolved the same way the Compose root resolves it: the mirror
     * this class itself keeps up to date, then the device, then the workspace.
     */
    private suspend fun uiLocale(): String = UiLocale.resolve(
        graph.prefs.uiLocale.first(),
        UiLocale.deviceTag(),
        graph.prefs.workspaceLocale.first(),
    )
}

File: apps/ios/Loonext/Features/Shell/RootViewModel.swift
import Foundation
import Observation

/// Mirrors the web's CompanyProvider bootstrap (and the Android
/// RootViewModel): session → GET /v1/me → resolve active company (persisted
/// pick or first membership) → route.
enum RootState {
    case loading
    case signedOut

    /// Signed in, zero memberships — workspace creation lives on web (checkout).
    case needsWorkspace(Me)

    /// Owner/admin with subscription_status incomplete — finish checkout on web.
    case needsCheckout(Me, companyId: String)

    /// #496/#314: this session has not satisfied two-factor.
    /// `enrolmentRequired` tells the gate which question to ask — a code from
    /// somebody who has a factor, or enrolment first for somebody the WORKSPACE
    /// is insisting on and who has none.
    case needsMfa(enrolmentRequired: Bool)

    case ready(Me, companyId: String)

    case failed(String)
}

/// #483: the waits before each retry of a bootstrap number list that failed.
///
/// GET /v1/numbers is the only source of the access-filtered list that decides
/// which per-number realtime topics this client joins, and when it fails the
/// socket opens with an EMPTY one — the company topic, and not a single
/// per-number topic. The reconnect observer in `start()` heals that on the next
/// re-JOIN, which on a healthy socket can be hours away, and since #484's
/// contract step those are hours of an inbox that never updates. Three tries
/// across ~17s turn one transient failure into a blink; going longer would only
/// race the re-JOIN heal that already exists. (Android's ladder, same numbers.)
let numberListRetryDelays: [Duration] = [.seconds(1), .seconds(4), .seconds(12)]

/// Run `attempt` after each wait in `delays` until one reports success.
///
/// Waits FIRST: a read that just failed fails the same way if it is repeated in
/// the same millisecond. `Task.isCancelled` is then checked explicitly because
/// `try?` swallows the cancellation error out of `Task.sleep` — without that
/// check a cancelled ladder would run its remaining attempts back to back,
/// reading /v1/numbers for a company the app has already moved off.
///
/// `delays` is a parameter, defaulted to the production ladder, so a test can
/// describe this schedule in milliseconds: XCTest has no virtual clock to skip
/// the seventeen seconds the way kotlinx-coroutines-test does for Android's twin.
func retryNumberList(
    delays: [Duration] = numberListRetryDelays,
    attempt: @Sendable () async -> Bool
) async -> Bool {
    for delay in delays {
        try? await Task.sleep(for: delay)
        if Task.isCancelled { return false }
        if await attempt() { return true }
    }
    return false
}

@MainActor
@Observable
final class RootViewModel {
    private(set) var state: RootState = .loading

    private let graph: AppGraph
    private var started = false

    /// Bumped by every realtime number re-derive; see `resubscribeNumbers`.
    private var numbersGeneration = 0

    /// #483: the in-flight `retryNumberList`, so bootstraps cannot stack them.
    private var numberListRetry: Task<Void, Never>?

    /**
     #289: the pending background drop, so a quick app-switch cancels it rather
     than stacking a second one behind it.
     */
    private var realtimeDrop: Task<Void, Never>?

    /**
     #289: whether the socket is down because the app was backgrounded, as
     opposed to never having been up.

     Needed because becoming active fires on every return to the app, and
     reconnecting a socket that is already connected would tear down a live one
     for nothing.
     */
    private var droppedForBackground = false

    init(graph: AppGraph) {
        self.graph = graph
    }

    /// Idempotent; called once from the root view's `.task`.
    func start() {
        guard !started else { return }
        started = true

        // A dead refresh token anywhere lands back on login. Each closure
        // boundary upgrades `weak self` (a captured var) to a strong `let`
        // before the next @Sendable capture — Swift 6 forbids referencing the
        // var from nested concurrent code (CI run 6).
        Task { [weak self] in
            guard let self else { return }
            await self.graph.api.setSignedOutHandler { [weak self] in
                guard let self else { return }
                Task { @MainActor in
                    await self.graph.realtime.disconnect()
                    self.state = .signedOut
                }
            }
        }

        // Session appearing/disappearing drives everything.
        Task { [weak self] in
            guard let changes = self?.graph.sessionStore.changes else { return }
            for await session in changes {
                guard let self else { return }
                if session == nil {
                    await self.graph.realtime.disconnect()
                    self.state = .signedOut
                } else if self.isSignedOutOrLoading {
                    await self.bootstrap()
                }
            }
        }

        // Initial route from the persisted session.
        Task { [weak self] in
            guard let self else { return }
            if self.graph.sessionStore.current() == nil {
                self.state = .signedOut
            } else {
                await self.bootstrap()
            }
        }

        // #480: number access changed somewhere in this company — re-derive the
        // subscription set, not just the screens' data.
        //
        // Realtime authorization is a JOIN-TIME handshake: the topic policy runs
        // on `phx_join` and on a pushed token, never per broadcast. So a member
        // who just lost a number keeps receiving that number's events for up to
        // an hour unless the client LEAVES the channel, which is what asking the
        // server for the list again and re-subscribing does. The payload names
        // only the company, so nobody can tell whether they were the subject —
        // everyone re-derives and the server's answer decides.
        Task { [weak self] in
            guard let realtime = self?.graph.realtime else { return }
            for await event in await realtime.events()
                where event.event == "access.changed" {
                guard let self else { return }
                await self.resubscribeNumbers()
            }
        }

        // Broadcasts are not replayed, so an `access.changed` that landed while
        // the socket was down is simply gone — re-derive on every re-JOIN too.
        // This is also the LAST resort for a number list that failed to load at
        // bootstrap: it heals that too, but a re-JOIN can be hours away on a
        // healthy socket, which is why `numberListRetry` gets there first (#483).
        // It overlaps the observer above (an `access.changed` also fires the
        // reconnect signal) and that is fine: the newest answer wins and an
        // unchanged set costs nothing.
        Task { [weak self] in
            guard let realtime = self?.graph.realtime else { return }
            for await _ in await realtime.reconnected() {
                guard let self else { return }
                await self.resubscribeNumbers()
            }
        }
    }

    private var isSignedOutOrLoading: Bool {
        switch state {
        case .signedOut, .loading: true
        default: false
        }
    }

    func retry() {
        state = .loading
        Task { await self.bootstrap() }
    }

    func switchWorkspace(_ companyId: String) {
        graph.prefs.setActiveCompany(companyId)
        Task { await self.bootstrap() }
    }

    func signOut() {
        // The session-changes stream observes the clear and routes to login.
        Task {
            // Best-effort device push-token delete BEFORE the session clears —
            // the DELETE needs the bearer (#151); failure never blocks sign-out.
            await PushCoordinator.shared.ensureRegistrar(api: graph.api).unregister()
            // Revoke the session SERVER-side, in the same "needs the bearer" slot
            // and for a sharper reason. `AuthManager.signOut` ends the GoTrue
            // session and nothing else, so `user_sessions.revoked_at` stayed null
            // and the softphone credential was never swept — and because
            // authorization is `revoked_at is null`, this phone's access token kept
            // full read and send for the rest of its life after Sign out.
            //
            // Best-effort like the line above: nobody may be trapped in an account
            // because the network blipped on the way out. A failure means the token
            // expires on its own instead of being cut short, which is where every
            // sign-out already was.
            _ = try? await SettingsRepository(
                api: graph.api,
                sessionStore: graph.sessionStore
            ).revokeThisSession()
            await self.graph.authManager.signOut()
        }
    }


    // MARK: - #289 the socket a backgrounded phone should not be holding

    /**
     The app went to the background.

     Both apps connected on sign-in and disconnected on sign-out, so a phone in
     a pocket kept a WebSocket alive and sent a heartbeat every 25 seconds all
     day. The bytes are nothing; the RADIO is the cost — on LTE each
     transmission holds the modem in a high-power state for seconds afterwards,
     so a packet every 25 seconds never lets it sleep.

     Nothing is lost by dropping it: push carries every message, task and — the
     unforgivable one — every incoming call, which arrives through PushKit and
     never by a socket frame.
     */
    func appDidEnterBackground() {
        realtimeDrop?.cancel()
        guard let wait = RealtimeLifecycle.dropDelayMs(
            foreground: false,
            backgroundedForMs: 0,
            callActive: Self.callActive()
        ) else { return }
        realtimeDrop = Task { [weak self] in
            try? await Task.sleep(for: .milliseconds(wait))
            guard !Task.isCancelled, let self else { return }
            // Asked again after the wait: a call can start inside the grace
            // window — a push wakes the app, it rings, and the person answers
            // from the lock screen without the app ever coming forward.
            guard !Self.callActive() else { return }
            self.droppedForBackground = true
            await self.graph.realtime.disconnect()
        }
    }

    /** The app came back. Cancel a pending drop, or undo one that happened. */
    func appDidBecomeActive() {
        realtimeDrop?.cancel()
        realtimeDrop = nil
        guard droppedForBackground else { return }
        droppedForBackground = false
        // Not `connect` directly: the token may have expired while the app was
        // away and the number list may have changed. Bootstrap is the path that
        // re-derives both, and it is what a cold start runs anyway.
        Task { [weak self] in await self?.bootstrap() }
    }

    /**
     Is a call live on this device right now?

     Read through `peek()` rather than `get(graph:)` because the realtime
     lifecycle must not be the thing that forces a softphone into existence — a
     phone that has never made a call should not build one to find out it has no
     call. No manager means no call, which is the honest answer and the safe
     one: the socket drops, and an incoming call still arrives by push.
     */
    private static func callActive() -> Bool {
        CallsManager.peek()?.state.activeId != nil
    }

    private func bootstrap() async {
        do {
            let me = try await graph.meApi.me()
            guard !me.memberships.isEmpty else {
                // #228: their own setting still decides, even with no workspace
                // to fall back to — the hand-off screen is a screen.
                UiLocaleStore.shared.apply(user: me.locale, company: nil)
                state = .needsWorkspace(me)
                return
            }
            let stored = graph.prefs.activeCompanyId
            let membership = me.memberships.first { $0.company_id == stored } ?? me.memberships[0]
            graph.prefs.setActiveCompany(membership.company_id)
            // #228: the whole locale chain, settled before anything paints —
            // their own setting, then their phone (which the store holds), then
            // this workspace's. Applied on EVERY bootstrap, so a language
            // changed on their laptop corrects here without a sign-out, and a
            // workspace switch moves the last step to the new one.
            //
            // Deliberately not cleared on sign-out: the member's own choice is
            // theirs and outlives the session, and the workspace's is only ever
            // consulted when the other two say nothing.
            UiLocaleStore.shared.apply(user: me.locale, company: membership.locale)

            let incomplete = membership.subscription_status == SubscriptionStatus.incomplete ||
                membership.subscription_status == SubscriptionStatus.incompleteExpired
            if incomplete, MemberRole.atLeast(membership.role, required: MemberRole.admin) {
                state = .needsCheckout(me, companyId: membership.company_id)
                return
            }

            // Connect realtime for the active workspace.
            //
            // #283: unless the realtime kill switch is off. This is the one
            // switch the server cannot enforce — the app holds its own Supabase
            // token and opens its own socket, so there is nothing for the
            // Worker to refuse. Not connecting leaves the app on its normal
            // refresh path: slower, never wrong.
            //
            // `!= false` rather than a truthiness check: an absent flag means
            // "no statement", which must read as ON.
            let realtimeAllowed = me.flags?["kill:realtime"] != false
            // #483: only ever true when the read was actually attempted AND failed.
            // An empty list is a real answer — a member restricted out of every
            // number joins the company topic alone and works — so it must not be
            // retried at; a failure wears the same shape by accident and must be.
            var numberListFailed = false
            if realtimeAllowed, let session = graph.sessionStore.current() {
                // #480: the per-number topics are joined with the company one, so
                // the list is awaited here rather than filled in afterwards — a
                // socket that opens with an incomplete subscription set has a gap
                // in it, and one small GET is the honest price of not having one.
                let numberIds = await visibleNumberIds(membership.company_id)
                numberListFailed = numberIds == nil
                await graph.realtime.connect(
                    companyId: membership.company_id,
                    numberIds: numberIds ?? [],
                    accessToken: session.accessToken,
                    // #302: the presence key. Nothing else on this socket says
                    // anything about who the viewer is.
                    userId: me.user_id
                )
            }
            // #496 — "I am able to login without any 2fa codes even though 2fa
            // is enabled." Asked POSITIVELY as well as caught below, because
            // there is a window where the server is not refusing yet: a Worker
            // deployed ahead of the migration reports no enrolment, and the app
            // should still ask. GET /v1/mfa is company-exempt, so it answers
            // even for a session every other route is refusing — which is
            // exactly the session that needs the answer.
            let mfa = try? await SettingsRepository(
                api: graph.api,
                sessionStore: graph.sessionStore
            ).mfa()
            // `aal` absent reads as aal1, matching the Worker's own reading of
            // the claim — the conservative direction, since the only thing that
            // turns it into aal2 is a factor this person actually presented.
            if let mfa, mfa.isEnrolled, (mfa.aal ?? "aal1") != "aal2" {
                state = .needsMfa(enrolmentRequired: false)
                return
            }

            state = .ready(me, companyId: membership.company_id)

            // #483: one transient failure of that GET must not cost a whole session
            // of per-number realtime. Retried OFF the bootstrap path — the shell is
            // already `.ready` above, so nothing here is waited on — and through
            // `resubscribeNumbers`, which is exactly the work the `access.changed`
            // and reconnect observers do.
            //
            // A ladder still pending from an earlier bootstrap is stale whatever
            // happened here: this pass has just read for the company it would have
            // read for.
            numberListRetry?.cancel()
            numberListRetry = nil
            if numberListFailed {
                numberListRetry = Task { [weak self] in
                    guard let self else { return }
                    // Strong `self` from here down: Swift 6 forbids referencing the
                    // captured weak var from the nested @Sendable closure (see the
                    // same upgrade in `start()`).
                    _ = await retryNumberList { await self.resubscribeNumbers() }
                }
            }
        } catch let error as ApiError {
            // #314/#496: both are walls with a route out, not failures.
            // Rendering them as "Couldn't load your workspace." — which is what
            // happened before, on both phones — is a lockout with no
            // explanation attached.
            switch error.code {
            case ApiErrorCode.unauthorized:
                state = .signedOut
            case ApiErrorCode.mfaChallengeRequired:
                state = .needsMfa(enrolmentRequired: false)
            case ApiErrorCode.mfaRequired:
                state = .needsMfa(enrolmentRequired: true)
            default:
                state = .failed(error.message)
            }
        } catch {
            // #228: the one sentence on this path that is ours rather than the
            // server's, so it is the one that has to be translated. Read off the
            // store rather than the environment — there is no view here, and
            // threading a locale down to one catch block would be worse than
            // asking the app's single answer for it.
            state = .failed(
                AppStrings.translate(
                    UiLocaleStore.shared.resolved,
                    "shell.bootstrapFailed"
                )
            )
        }
    }

    /// The ids of the numbers this member may see — one realtime topic each
    /// (#480). GET /v1/numbers is ALREADY access-filtered server-side (#106), so
    /// this asks rather than reasons: a client-side reading of the access rule
    /// would be the extra implementation D88 spent an issue removing.
    ///
    /// `nil` is a FAILED read, and telling it apart from an empty list is the
    /// whole point (#483): an empty list is a real answer for a member restricted
    /// out of every number, who joins the company topic alone and works normally,
    /// while a failure produces that same shape by accident. The caller degrades
    /// either way — the company topic alone is exactly how this behaved before
    /// per-number topics existed — but only the failure is worth retrying.
    private func visibleNumberIds(_ companyId: String) async -> [String]? {
        let page: Page<PhoneNumberSummary>? = try? await graph.api.get(
            "/v1/numbers",
            companyId: companyId
        )
        return page?.data.map(\.id)
    }

    /// Re-derive the number list and move the realtime subscription onto it.
    /// Only in `.ready`: any other state has no socket to reconcile.
    ///
    /// Two signals drive this and access can change twice in a row, so a slow
    /// response must not land after a newer one and re-subscribe to a set that
    /// has already been superseded — which, for a revoked number, would mean
    /// staying on its channel with nothing left to correct it.
    ///
    /// Reports whether a fresh list was actually applied, which is what the #483
    /// bootstrap ladder needs to know when it is standing in for a GET that
    /// failed. A superseded answer reports `false` too: a newer read owns the
    /// subscription set now, and if the ladder has rungs left, one more cheap GET
    /// beats betting the heal on a result we just discarded. The two observers in
    /// `start()` ignore it — they fire off a signal that will come again, so there
    /// is nothing for them to do about a `false`.
    @discardableResult
    private func resubscribeNumbers() async -> Bool {
        guard case .ready(_, let companyId) = state else { return false }
        numbersGeneration += 1
        let generation = numbersGeneration
        guard let numberIds = await visibleNumberIds(companyId) else { return false }
        guard generation == numbersGeneration else { return false }
        await graph.realtime.setNumbers(numberIds)
        return true
    }
}
